Detectors/MUON/MCH/Tracking/src/TrackFinder.cxx

```diff
@@ -0,0 +1,1668 @@
+      continue;
+    }
+
+    // look for cluster candidate in this DE
+    for (const auto& cluster : *de.second) {
+
+      // try to add the current cluster
+      if (!isCompatible(currentParam, cluster, paramAtCluster)) {
+        continue;
+      }
+
+      // duplicate the track and add the new cluster
+      itNewTrack = mTracks.emplace(itNewTrack, *itTrack);
+      print("followTrackInOverlapDE: duplicating candidate at position #", getTrackIndex(itNewTrack), " to add cluster ", cluster.getIdAsString());
+      itNewTrack->addParamAtCluster(paramAtCluster);
+
+      // tag the track as removable (if it is not already the case) if it is out of limits
+      if (!itNewTrack->isRemovable() && !isAcceptable(paramAtCluster)) {
+        itNewTrack->removable();
+      }
+    }
+  }
+
+  return (itNewTrack == itTrack) ? mTracks.end() : itNewTrack;
+}
+
+//_________________________________________________________________________________________________
+std::list<Track>::iterator TrackFinder::followTrackInChamber(std::list<Track>::iterator& itTrack,
+                                                             int chamber, int lastChamber, bool canSkip,
+                                                             std::unordered_map<int, std::unordered_set<uint32_t>>& excludedClusters)
+{
+  /// Follow the track candidate pointed to by "itTrack" to the given "chamber"
+  /// The tracking starts from the current parameters, which must have already been set
+  /// Look for compatible cluster(s), excluding those in the "excludedClusters" list, which
+  /// correspond to compatible clusters already associated to this candidate in a previous step
+  /// For each (pair of) cluster(s) found, continue the tracking to the next chamber, up to "lastChamber"
+  /// This is a recursive procedure. Once reaching the last requested chamber, every valid tracks found
+  /// are added before "itTrack" and the associated clusters from this chamber onward are attached to them
+  /// The method returns an iterator to the first new candidate, or mTracks.end() if none is found
+  /// Every compatible clusters found in the process are added to the "excludedClusters" list
+  /// The initial candidate "itTrack" is not modified, with the exception of its current parameters,
+  /// which are set to the parameters at "chamber" or invalidated in case of propagation issue
+
+  print("followTrackInChamber: follow track #", getTrackIndex(itTrack), " to chamber ", chamber + 1, " up to chamber ", lastChamber + 1);
+
+  // the current track parameters must be set at a different chamber and valid
+  if (!itTrack->areCurrentParamValid() || chamber == itTrack->getCurrentChamber()) {
+    return mTracks.end();
+  }
+
+  // determine the 2 planes or 4 half-planes of the chamber, ordered according to the direction of propagation
+  int plane[4] = {-1, -1, -1, -1};
```

Comment:
Hi @shahor02 ,

Indeed this can be improved. In fact with the present algorithm the direction of propagation is unique for each chamber so this part does not need to be generic and it can be fully determined at compilation time. I changed the code to use a constexpr.